### PR TITLE
Feature/burn unusable tokens

### PR DESCRIPTION
**Pull request by cducrest**
closes https://github.com/trustlines-protocol/merkle-drop/issues/23

### CODE REVIEW

**Comment by cducrest** [comment]
Need to fix a problem arising when attempting to burn tokens later than the decayduration

**Comment by cducrest** [comment]
Ok, I should burn some tokens on withdraw but I did not find an easy calculation for this, and the calculation I had initially was wrong since the amount to burn there is not `value - valueToSend` because some of it has potentially been burnt already in `burnUnusableTokens`.

I can make a new calculation or call `burnUnusableTokens()` at the beginning of withdraw. Do you have a recommendation?

**Comment by schmir-at-bb** [comment]
My recommendation is to have real accounting of the tokens burnt and withdrawn.  At any point in time you could compute the tokens which decayed from the sum of entitlements at the start minus the entitlements withdrawn. You could compare that to the number of tokens burnt via the burn function and burn the difference. (I haven't thought this through completely)

**Comment by schmir-at-bb** [comment]
I thought a bit more about this. I would try to account the following things:

- `entitlement_not_withdrawn`: The sum of all initial entitlements for all users who have not withdrawn their entitlement. If a user withdraws we reduce this value by the initial entitlement of the user.
- `burned_entitlement_not_withdrawn`: Sum of all burned values from the `entitlement_not_withdrawn`. When we burn value, we add the burned value here. When a user withdraws, we reduce this by the burned part of his initial entitlement. If we want to burn value, we can compute the difference between this value and the to be burned part of the `entitlement_not_withdrawn` computed from the beginning of the decay period and burn this difference. This way we keep rounding errors to a minimum.

Also, when a user withdraws, we always burn from the complete `entitlement_not_withdrawn`. This makes sure the users entitlement to be burned is already burned. There is no need to have the burning logic for a single user (though we do have to keep track of the accounting)

-  `entitlement_withdrawn`: The sum of all initial entitlements of all user who have withdrawn
- `burned_entitlement_withdrawn`: The burned part of all initial entitlements of all users who have withdrawn
- `transferred_entitlement_withdrawn`: The sum of all transferred values to users who have withdrawn

**Comment by schmir-at-bb** [comment]
The burn function can still be simplified.

I've opened a PR for this branch here: https://github.com/trustlines-protocol/merkle-drop/pull/30

File: contracts/contracts/DroppedToken.sol [hunk 1]
```diff
@@ -75,6 +84,10 @@ contract DroppedToken {
     }
 
     function burn(uint256 amount) public {
+        if (! burnLoopFlag) {
+            burnLoopFlag = true;
+            merkleDrop.burnUnusableTokens();
+        }
```
Comment (schmir-at-bb) [inline contracts/contracts/DroppedToken.sol]: I'm still not convinced that we need this flag. Also I don't see where it is being set to false, which makes me wonder why the tests even succeed.
Why do we need to burn the tokens in the merkleDrop if someone burns his tokens? Same Question for the transfer function.

File: contracts/contracts/MerkleDrop.sol [hunk 2]
```diff
@@ -77,11 +77,12 @@ contract MerkleDrop {
         }
 
         // The amount of tokens that should be held within the contract after burning
-        uint decayedRemainingValue = remainingValue - decay(remainingValue, now - decayStartTime, decayDurationInSeconds);
+        uint targetBalance = decayedEntitlementAtTime(remainingValue, now);
 
-        assert(decayedRemainingValue + spentTokens <= initialBalance);
+        assert(targetBalance + spentTokens <= initialBalance);
```
Comment (schmir-at-bb) [inline contracts/contracts/MerkleDrop.sol]: Please change this to
```
assert(targetBalance <= currentBalance)
```
and move the currentBalance assignment before this assert.

**Pull request merged**